File: KthSmallestInSortedMatrix.py
#All TC passed on leetcode

class Solution:
    def kthSmallest(self, matrix: List[List[int]], k: int) -> int:

        #Here we perform binary search on nums b/w smallest and largest num of matrix. 
        #Then we count the nums in matrix that are lesser than mid, if count is greater than k then we 
        #shift our search space to left else we shift our search space to right.
        #Time complexity - O(n.logH) where n - number of rows, and H is the numbers b/w lowest and highest number in the matrix 
        #Space complexity - O(1)
        n = len(matrix)
        # Set the range of the search to be between the smallest element
        # and the largest element in the matrix
        left, right = matrix[0][0], matrix[n - 1][n - 1]

        # Binary search for the kth smallest element
        while left <= right:
            mid = (left + right) // 2
            count = 0
            j = n - 1
            # Count the number of elements in the matrix that are less than or equal to mid
            for i in range(n):
                while j >= 0 and matrix[i][j] > mid:
                    j -= 1
                count += j + 1
            # Update the range of the search
           
            if count < k:
                left = mid + 1
            else:
                right = mid - 1

        return left

#--------------------------------------OR----------------------------------------------
        
        #Time complexity - O(X + KlogX) where X = min(k, rows)
        #Space complexity - O(X) - heap size
        minHeap = []
        rows = len(matrix)
        cols = len(matrix[0])
        
        for r in range(min(k,rows)):
            minHeap.append([matrix[r][0], r, 0])
            
        heapq.heapify(minHeap)
        
        while k:
            k-=1
            val, r, c = heapq.heappop(minHeap)
            if k==0:
                return val
            if c+1 < cols:
                heapq.heappush(minHeap, [matrix[r][c+1], r, c+1])
            
        
        

File: kthSmallestInSortedMatrix.java
/* Using binary search, and taking the min and max element in the matrix, we do the following
 1. Get the middle element, 
 find the total number of elements which are less than or equal to this number in the matrix using function count_elem_less_than_num. 
The advantage is : that kth smallest numb which we are  trying to find is going to be at the point where we get the count 
from the function count_elem_less_than_num = our target.  
 If the total number of elements are more than what you searched for in matrix, 
it means you need to move your search window to the left to accomodate less elements. 
so the new search will be : start remains the same.end will change to mid. 
Here we are not changing to mid -1 because there can be a case where mid is the actual point where we get our target count elements. 
 If total number of elements is less than what you searched for in matrix, 
it means you need to increase your search window
 to do that... change start to mid + 1 and end remains the same so that we can get a higher window to search for
*/
// Time Complexity: O(klogk)    (where k = mn in the worst case)
    // Space Complexity: O(1)


class Solution {
    public int kthSmallest(int[][] matrix, int k) {
        int n = matrix.length;
        
        int low = matrix[0][0];
        int high = matrix[n-1][n-1];
        while(low < high){
            int mid = low+ (high -low)/2;
            int count = lessOrEqual(matrix, mid, n);
            if(count < k){
                low = mid +1;
            }
            else {
                high = mid;
            }
            
        }
        return low;
    }
    
        
        public int lessOrEqual(int[][]matrix, int mid, int n){
            int result =0;
            int j= n-1;
            for(int i=0;i<matrix.length;i++){
                while( j>=0 && matrix[i][j] > mid)
                    j--;
                    result += j+1;
               
            }
            return result;
        }

               
    
}


/*APPROACH 2:uSING min Heap */ 
// Time Complexity: O(klogk)    (where k = mn in the worst case)
    // Space Complexity: O(k)

class KthSmallestElementInSortedMatrix {
    
    
    
    public int kthSmallest(int[][] matrix, int k) {
        // Edge Case Checking
        if(matrix == null || matrix.length == 0 || matrix[0].length == 0 || k <= 0)
            return -1;
        
        int n = matrix.length;
        int m = matrix[0].length;
        // Creating a max heap
        PriorityQueue<Integer> maxHeap = new PriorityQueue<>((a, b) -> b - a);

        // Traverse through the matrix and add elements to the heap
        for(int i = 0; i < n; i++){
            for(int j = 0; j < m; j++){
                maxHeap.add(matrix[i][j]);
                // If the heap size goes greater than k - remove the root element
                if(maxHeap.size() > k)
                    maxHeap.poll();
            }
        }

        // After the traversal - the root element will be the kth smallest element
        return maxHeap.peek();
    }
}